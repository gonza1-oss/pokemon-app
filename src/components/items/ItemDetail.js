import React from 'react';
import { Redirect, Link } from "react-router-dom";

import getItems from '../../services/getItem';
import { getLoadingSpinner, getError } from '../../services/common';

function ItemDetail ({match}) {
  var evolutionsList = null;

  // graphql call to get the pokemon passing the id
  if (!match) return getError();
  const { loading, error, data } = getItems(match.params.item);
  if (loading) return getLoadingSpinner();
  if (error) return getError();
  if (data.pokemon === null) return <Redirect to={{ pathname: "/" }} />;

  const { image, name, number, types, maxHP, maxCP, evolutions } = data.pokemon;

  // render evolutions if exists
  if (evolutions) {
      evolutionsList = evolutions.map(({ id, name, image }) => (
      <div className="ui card" key={id}>
        <div className="meta"><h5>{name}</h5></div>
        <div className="image" style={{ backgroundImage: `url('${image}')`, backgroundColor: 'white', backgroundSize: 'contain', backgroundPosition: 'center', backgroundRepeat: 'no-repeat', height: '5rem' }} ></div>
        <div className="ui bottom attached button">
          <Link to={`/${id}`}>
            <i className="ellipsis vertical icon"></i>
            More
          </Link>
        </div>
      </div>
    ));
  }

  //render the Pokemon Detail page
 return <div className="ui container">

          {/* Back button */}
          <div className="ui container">
            <Link to={'/'} className="ui left floated labeled icon button">
              <i className="left arrow icon"></i>Back
            </Link>
          </div>

          {/* pokemon detail box */}
          <div className="ui container">
            <div className="ui relaxed divided items">
              <div className="item">

                {/* pokemon image box */}
                <div className="ui big image" >
                  <img alt={name} src={image} />
                </div>
                {/* pokemon description box */}
                <div className="content">
                  <h1>{name}</h1>
                  <div className="meta">
                    <span className="date">{number}</span>
                  </div>
                  <div className="description">
                    <p><b>Max HP:</b> {maxHP}</p>
                    <p><b>Max CP:</b> {maxCP}</p>
                    { types && <p><b>Type:</b> {types.join(', ')}</p>}
                  </div>

                  {/* evolution box */}
                  { evolutionsList &&
                    <div className="description">
                      <h4>Evolutions:</h4>
                      <div className="ui two doubling cards">{evolutionsList}</div>
                    </div>
                  }

                </div>
              </div>
            </div>
          </div>
        </div>;
}

export default ItemDetail;
