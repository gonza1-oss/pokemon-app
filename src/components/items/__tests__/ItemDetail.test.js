import React from 'react';
import ReactDOM from 'react-dom';
import { gql } from 'apollo-boost';
import { MockedProvider } from '@apollo/react-testing';
import { act, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import ItemDetail from '../ItemDetail';

const match = { params: { item: 'UG9rZW1vbjowMDE=' } };

const QUERY = gql`
{
  pokemon(id: "UG9rZW1vbjowMDE=") {
    id
    number
    name
    maxCP
    maxHP
    image
    types
    evolutions {
      id
      number
      name
      maxCP
      maxHP
      image
      types
    }
  }
}`;

const MOCKS = [
  {
    request: {
      query: QUERY,
    },
    result: {
      data: {
        pokemon: {
          id: "UG9rZW1vbjowMDE=",
          number: "001",
          name: "Bulbasaur",
          maxCP: 951,
          maxHP: 1071,
          image: "https://img.pokemondb.net/artwork/bulbasaur.jpg",
          types: [
            "Grass",
            "Poison"
          ],
          evolutions: [{
            id: "UG9rZW1vbjowMDI=",
            number: "002",
            name: "Ivysaur",
            maxCP: 1483,
            maxHP: 1632,
            image: "https://img.pokemondb.net/artwork/ivysaur.jpg",
            types: [
              "Grass",
              "Poison"
            ]
          }],
        },
      },
    },
  },
];

const MOCKS_NULL = [
  {
    request: {
      query: QUERY,
    },
    result: {
      data: {
        pokemon: null
      }
    }
  },
];

const MOCKS_ERROR = [
  {
    request: {
      query: QUERY,
    },
    result: null
  },
];

async function wait(ms = 0) {
  await act(() => {
    return new Promise(resolve => {
      setTimeout(resolve, ms);
    });
  });
}

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MockedProvider mocks={[]}><ItemDetail match={match} /></MockedProvider>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders pokemon item', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS}>
        <ItemDetail match={match} />
      </MockedProvider>
    </BrowserRouter>
  );

  await wait();

  expect(container.textContent).toMatchSnapshot();

  expect(container.textContent).toMatch('Bulbasaur');
  expect(container.textContent).toMatch('Type: Grass, Poison');

});

it('renders pokemon evolutions', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS}>
        <ItemDetail match={match} />
      </MockedProvider>
    </BrowserRouter>
  );

  await wait();

  expect(container.textContent).toMatch('Evolutions:Ivysaur');

});

it('renders error', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS_ERROR}>
        <ItemDetail match={match} />
      </MockedProvider>
    </BrowserRouter>
  );

  await wait(0);

  expect(container.textContent).toBe('Error :( ');

});

it('renders empty pokemon list', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS_NULL}>
        <ItemDetail match={match} />
      </MockedProvider>
    </BrowserRouter>
  );

  await wait(0);

  expect(container.textContent).toBe('');

});



// it('renders', async () => {
//
//   let root;
//
//   // act(() => {
//
//   root = create(
//       <MockedProvider mocks={[]}>
//         <ItemList />
//       </MockedProvider>
//     );
//
//     const tree = root.toJSON();
//
//     expect(tree).toMatchSnapshot();
//     await wait();
//     expect(tree).toContain('div');
//     //expect(root.textContent).toBe('Loading');
//
//   // })
//   await wait();
//
//   expect(root.toJSON()).toMatchSnapshot();
//
// });








// it('should render loading state initially', () => {
//   const component = create(
//     <MockedProvider mocks={[]}>
//       <ItemList />
//     </MockedProvider>,
//   );
//
//   const tree = component.toJSON();
//   expect(tree.children).toContain('Loading');
// });
