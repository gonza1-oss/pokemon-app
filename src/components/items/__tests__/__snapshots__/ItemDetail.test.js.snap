// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders pokemon item 1`] = `"BackBulbasaur001Max HP: 1071Max CP: 951Type: Grass, PoisonEvolutions:IvysaurMore"`;
